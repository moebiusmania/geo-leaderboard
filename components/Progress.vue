<script setup lang="ts">
const props = defineProps<{
  levels: number;
  current: number;
  cap: number;
  score: number;
  progress: number;
}>();

const circles = Array.from({ length: props.levels }, (_, i) => i + 1);
</script>

<template>
  <div class="progress-container">
    <div class="progress-bar">
      <div class="progress-fill" :style="{ width: `${progress}%` }" />
      <div class="circles">
        <div
          v-for="circle in circles"
          :key="circle"
          class="circle"
          :class="{ active: circle <= current }"
        />
      </div>
    </div>
  </div>
</template>

<style scoped>
.progress-container {
  width: 100%;
  padding: 20px 0;

  & .progress-bar {
    position: relative;
    height: 6px;
    background: var(--text);
    border-radius: 2px;
  }

  & .progress-fill {
    position: absolute;
    height: 100%;
    background: var(--primary);
    border-radius: 2px;
    transition: width 0.3s ease;
  }

  & .circles {
    position: absolute;
    top: 50%;
    width: 100%;
    display: flex;
    justify-content: space-between;
    transform: translateY(-50%);
  }

  & .circle {
    --size: 20px;

    width: var(--size);
    height: var(--size);
    background: var(--text);
    border-radius: 50%;
    transition: background-color 0.3s ease;
  }

  & .circle.active {
    background: var(--primary);
  }
}
</style>
