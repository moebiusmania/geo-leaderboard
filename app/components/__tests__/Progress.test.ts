import { describe, it, expect } from "vitest";
import { mount } from "@vue/test-utils";
import Progress from "../Progress.vue";

describe("Progress", () => {
  const defaultProps = {
    levels: 5,
    current: 3,
    cap: 100,
    score: 75,
    progress: 75,
  };

  it("renders properly", () => {
    const wrapper = mount(Progress, {
      props: defaultProps,
    });
    expect(wrapper.exists()).toBe(true);
  });

  it("renders correct number of circles", () => {
    const wrapper = mount(Progress, {
      props: defaultProps,
    });
    const circles = wrapper.findAll(".circle");
    expect(circles).toHaveLength(defaultProps.levels);
  });

  it("marks correct number of circles as active", () => {
    const wrapper = mount(Progress, {
      props: defaultProps,
    });
    const activeCircles = wrapper.findAll(".circle.active");
    expect(activeCircles).toHaveLength(defaultProps.current);
  });

  it("sets progress bar width correctly", () => {
    const wrapper = mount(Progress, {
      props: defaultProps,
    });
    const progressFill = wrapper.find(".progress-fill");
    expect(progressFill.attributes("style")).toBe("width: 75%;");
  });

  it("matches snapshot", () => {
    const wrapper = mount(Progress, {
      props: defaultProps,
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("updates when props change", async () => {
    const wrapper = mount(Progress, {
      props: defaultProps,
    });

    await wrapper.setProps({
      current: 4,
      progress: 80,
    });

    const activeCircles = wrapper.findAll(".circle.active");
    expect(activeCircles).toHaveLength(4);

    const progressFill = wrapper.find(".progress-fill");
    expect(progressFill.attributes("style")).toBe("width: 80%;");
  });
});
