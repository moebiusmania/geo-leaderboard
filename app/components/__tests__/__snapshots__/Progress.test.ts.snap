// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Progress > matches snapshot 1`] = `
"<div data-v-0b11ee78="" class="progress-container">
  <div data-v-0b11ee78="" class="progress-bar">
    <div data-v-0b11ee78="" class="progress-fill" style="width: 75%;"></div>
    <div data-v-0b11ee78="" class="circles">
      <div data-v-0b11ee78="" class="circle active"></div>
      <div data-v-0b11ee78="" class="circle active"></div>
      <div data-v-0b11ee78="" class="circle active"></div>
      <div data-v-0b11ee78="" class="circle"></div>
      <div data-v-0b11ee78="" class="circle"></div>
    </div>
  </div>
</div>"
`;
